### DISCUSSION TITLE
General SwitchBot devices status decoders

### DISCUSSION THREAD

**Discussion by DigiH** [theengs/decoder#95] [New devices/things integration]
Discussion about various SwitchBot devices' status implementations and testing of such …

**Comment by DigiH** [comment]
SwitchBot Curtain status decoder ready to test with

`decoder = https://github.com/DigiH/decoder#switchbot-curtain`

in platformio.ini

Would be great to get some feedback, especially if 

`motionstate` should report back the current strings `"stationary"`/`"movement"`

or if it's nicer to rename it and make it a blooean

`motion` should report back the current strings `false`/`true`

**Comment by 1technophile** [comment]
For the Switchbot W2301500 thermometer and hygrometer, here are some sample messages:
```
{"id":"E0:HH:HH:HH:HH:CE","mac_type":1,"manufacturerdata":"6909e02122c12fce060208993b","rssi":-63,"servicedata":"6900ba08993b","servicedatauuid":"0xfd3d"} 25.8C 56%
{"id":"E0:HH:HH:HH:HH:CE","mac_type":1,"manufacturerdata":"6909e02122c12fce0702089938","rssi":-56,"servicedata":"6900ba089938","servicedatauuid":"0xfd3d"} 25.8C 56%
{"id":"E0:HH:HH:HH:HH:CE","mac_type":1,"manufacturerdata":"6909e02122c12fce0b02009ab8","rssi":-51,"servicedata":"6900ba009ab8","servicedatauuid":"0xfd3d"} 78.8F 56% (set in Farenheit)
```

**Reply by DigiH** [reply]
Thanks a lot @1technophile, sample messages like that were missing for me to advance on this.

Improvements submitted

https://github.com/DigiH/decoder/tree/switchbot-meter

with your test cases added with the first humidity correction.

Also @h2zero, with the more complex decoder I had to increase the **m_docMax** to 9165.

As previously discussed, an implementation of a direct binary bit condition testing would greatly reduce the current multiple full hex digit OR comparisons.

My test implementation of such

https://github.com/theengs/decoder/compare/development...DigiH:hex-binary-test

does work great and doesn't really cause any overhead with the **bitset** include, but you said that you'd prefer a different solution.

Any chance we could get a  direct binary bit condition testing? This would greatly reduce many of the current implemented decoders to save lots of space and the need to increase **m_docMax**

Also when implementing the decimal part of the temperature for this decoder I tried to use

```
".cal":{
         "decoder":["value_from_hex_data", "servicedata", 6, 2, false, false],
         " "post_proc":["/", 10]
      },

…

      "_tempc":{
         "condition":["servicedata", 8, "8", "|", "servicedata", 8, "9", "|", "servicedata", 8, "a", "|", "servicedata", 8, "b", "|", "servicedata", 8, "c", "|", "servicedata", 8, "d", "|", "servicedata", 8, "e", "|", "servicedata", 10, "f"],
         "decoder":["value_from_hex_data", "servicedata", 8, 2, true, false],
         "post_proc":["+", ".cal", "-", 128]
      }, 
```

but either .cal cannot be a decimal, or adding a decimal in another post_proc is not possible!?!

So I ended up with a more convoluted
```

".cal":{
         "decoder":["value_from_hex_data", "servicedata", 6, 2, false, false]
      },

…

      "_tempc":{
         "condition":["servicedata", 8, "8", "|", "servicedata", 8, "9", "|", "servicedata", 8, "a", "|", "servicedata", 8, "b", "|", "servicedata", 8, "c", "|", "servicedata", 8, "d", "|", "servicedata", 8, "e", "|", "servicedata", 10, "f"],
         "decoder":["value_from_hex_data", "servicedata", 8, 2, true, false],
         "post_proc":["*", 10, "+", ".cal", "/", 10, "-", 128]
      },
```
which works fine, but I just wanted to mention the stumbling block above.

With the **(set in Fahrenheit)** case in your example above, Florian, I was already wondering how this would pan out, and if an addition **tempf** would have to be defined in the decoder which would then be converted into Celsius, but from the test cases at least it seems to be working as is and the Fahrenheit bit it only set for the app to know how to display the temperature. Maybe if you can try and do more tests with **(set in Fahrenheit)** just to confirm this.

Thanks

BTW. @1technophile, I fully understand the wish to mask the MAC addresses, I do the same whenever possible, but also manufacturerdata, index 4-15 ;)

**Reply by h2zero** [reply]
Seems I wrote my previous reply in the wrong box, oops. I forgot to reply to the post processing issue, so I'll do that here. Having a post process for the calculation value is not something that was considered when implemented but I can certainly see the value of it. I'll look to add this shortly, should be a minor adjustment. 

Update: This should work as is actually, the post process check is made before the value is assigned to the calculation value variable. I think the issue might be your decoder having a stray quotation: `" "post_proc":["/", 10]`

**Reply by DigiH** [reply]
Sorry for the confusion, the stray extra quotation mark above was actually just a copy and paste error of mine when recreating what I initially tried ;)

Just tested again with

```
".cal":{
         "decoder":["value_from_hex_data", "servicedata", 6, 2, false, false],
         "post_proc":["/", 10]
      },
      "tempc":{
         "condition":["servicedata", 8, "0", "|", "servicedata", 8, "1", "|", "servicedata", 8, "2", "|", "servicedata", 8, "3", "|", "servicedata", 8, "4", "|", "servicedata", 8, "5", "|", "servicedata", 10, "6", "|", "servicedata", 10, "7"],
         "decoder":["value_from_hex_data", "servicedata", 8, 2, true, false],
         "post_proc":["+", ".cal", "*", -1]
      },
      "_tempc":{
         "condition":["servicedata", 8, "8", "|", "servicedata", 8, "9", "|", "servicedata", 8, "a", "|", "servicedata", 8, "b", "|", "servicedata", 8, "c", "|", "servicedata", 8, "d", "|", "servicedata", 8, "e", "|", "servicedata", 10, "f"],
         "decoder":["value_from_hex_data", "servicedata", 8, 2, true, false],
         "post_proc":["+", ".cal", "-", 128]
      },
```

and the temperatures are now without the decimal value, i.e. without "+", ".cal"

```
1: Expected: {"brand":"SwitchBot","model":"Meter (Plus)","model_id":"SWITCHBOT-METER","tempc":25.8,"tempf":78.44,"hum":59,"batt":58}
1: Got JSON: {"brand":"SwitchBot","model":"Meter (Plus)","model_id":"SWITCHBOT-METER","tempc":25,"tempf":77,"hum":59,"batt":58}
```

Changing the .cal post_proc to

```
".cal":{
         "decoder":["value_from_hex_data", "servicedata", 6, 2, false, false],
         "post_proc":["/", 1]
      },

```

however makes it work, obviously giving wrong temperatures, but making it clearer that it is not the the post_proc in the .cal section, but the decimal that's an issue when using "+", ".cal" in the subsequent temperature post_procs ;)

**Reply by DigiH** [reply]
Tested with some more variations. The temperature post_procs, and any post_procs in general I assume, can only take integers, but not any floats as arguments.

**Reply by h2zero** [reply]
That makes sense, try changing this to a `double` https://github.com/theengs/decoder/blob/988bc6afa82631004f28bf7fa836da87c8571edf/src/decoder.cpp#L414 and remove the cast here https://github.com/theengs/decoder/blob/988bc6afa82631004f28bf7fa836da87c8571edf/src/decoder.cpp#L505

**Reply by DigiH** [reply]
Thanks @h2zero, working with the double now to have the shorter less convoluted temperature post_procs.

@1technophile would it be possible for you to also do some tests with sub-zero temperatures? I have amended one of the test cases and it seems fine, but an actual test with a device would be great. That is only if the SwitchBot Meter is fine in a freezer for a few minutes ;)

Thanks

**Reply by 1technophile** [reply]
@DigiH of course. I will do and let you know

**Reply by 1technophile** [reply]
Great, works perfect with the device, even in negative temperature:
```
{"id":"E0:21:22:C1:2F:CE","mac_type":1,"manufacturerdata":"6909e02122c12fce7d0e09043d","rssi":-49,"servicedata":"69c0ba09043d","servicedatauuid":"0xfd3d","brand":"SwitchBot","model":"Meter (Plus)","model_id":"SWITCHBOT-METER","tempc":-4.9,"tempf":23.18,"hum":61,"batt":58}
```

**Reply by 1technophile** [reply]
@DigiH for the model_id should we use `W2301500`, this is written on the back of the device.

**Reply by DigiH** [reply]
@1technophile Agreed, something like that would be better, but it seems that the SwitchBot Meter Plus has different model IDs (W2301500, W2301501,W2301502, W2301503, W2301504, W2301505), likely in different countries, and also wondering how to define the older SwitchBot Meter which also has "TH S1, TH J1, TH K1, TH E1, TH U1, TH C1"

https://support.switch-bot.com/hc/en-us/articles/5984263540631-Declaration-of-Conformity-Meter-Plus-

https://support.switch-bot.com/hc/en-us/articles/5984274786967-Declaration-of-Conformity-Meter-

`"model_id":"TH S-J-KE-U-C1/W230150-0-1-2-3-4-5",`

or probably

`"model_id":"TH *1/W230150*",`

I don't want to be awkward, but probably best not to take a model number restricted to one country/hardware revision ;)

Thanks

**Reply by 1technophile** [reply]
Good catch. Maybe we keep it like you have `SWITCHBOT-METER` and add the different models in the .md

**Reply by DigiH** [reply]
I've just tried with the latter suggestion above and I think it doesn't look bad

https://github.com/DigiH/decoder/commit/77f67c534ae0209e2a18e73bafe56f5814556313

unless you think the asterisk * might cause issues on some platform, an alternative would be

"model_id":"TH X1/W230150X",

or do you think it's best as you said above, revert to what I had before and list the modelIDs in the .md?

Also thanks for testing the sub-zero temperature ;) once we decide on the modelID this is good to go, and one of the SwitchBot inclusions which doesn't really require any WRITE command integration.

Thanks

**Reply by 1technophile** [reply]
Thanks,

Let's just remove the ampersand, add the list of models in the .md and we are good to go

**Reply by DigiH** [reply]
Hi @1technophile, naming the SwitchBot modelIDs in the .md files with an asterisk (*) sort of backfired

https://decoder.theengs.io/devices/SBMT.html

where it should have been
`
TH *1 (W230150*)`

now the two asterisks are invisible, but making the type in between them _italic_ ;)

So maybe really best to replace the asterisks with "X"s.

```
TH X1 (W230150X)
X1
```

And this could also then be presented as modelIDs in the JSON withouth any issues.

What do you think?

Thanks

**Reply by 1technophile** [reply]
> So maybe really best to replace the asterisks with "X"s.

Yes that works.

**Comment by h2zero** [comment]
@DigiH Yes the bin bit test is certainly a good addition. There is nothing wrong with your code, just that for such a simple case it is overkill 😄. 

Here is another version for you to try:
```
if (data_src) {
  size_t cond_len = strlen(prop_condition[i + 2 + inverse].as<const char*>());
  if (strstr((const char*)prop_condition[i + 2], "binbit") != nullptr) {
    char ch = *(data_src + prop_condition[i + 1].as<int>());
    uint8_t data = 0;
    if (ch >= '0' && ch <= '9')
      data = ch - '0';
    else if (ch >= 'a' && ch <= 'f')
      data = 10 + (ch - 'a');
    else if (ch >= 'A' && ch <= 'F')
      data = 10 + (ch - 'F');
    else
      return false;
  
    uint8_t shift = prop_condition[i + 3].as<uint8_t>();
    uint8_t val = prop_condition[i + 4].as<uint8_t>();
    if (((data >> shift) & 0x01) == val) {
      cond_met = true;
    }
    i += 2;
  } else if (!strncmp(&data_src[prop_condition[i + 1].as<int>()],
```

Also, maybe "binbit" should be just "bit"?

**Reply by DigiH** [reply]
Thanks @h2zero @1technophile , I added the bit evaluation condition and converted the new SwitchBot Meter and the existing SwitchBot S1 decoder, saving quite a bit of real estate and leaving m_docMax untouched.

https://github.com/DigiH/decoder/commit/ffdbd6e36eeeafbb9425599f718f5ca7bcdfdcac

This will also be a prerequisite for most other upcoming SwitchBot decoders and a few existing decoders will also benefit from an adjustment.

**Comment by DigiH** [comment]
**SwitchBot Motion Sensor** initial status decoder ready for testing with

decoder = https://github.com/DigiH/decoder/tree/switchbot-motionsensor

in platformio.ini

Would be great to get some feedback about

• are`LED state` and `IoT state` really required to be reported?
• is `scope tested` fine wiht only showing up when `false`?

Thanks

**Reply by h2zero** [reply]
Wow that thing is huge lol, trimming anything unnecessary would be the best practice.

**Reply by DigiH** [reply]
As I previously wondered, who would have thought that a simple motion sensor would be so overloaded with possible properties ;) you really wouldn't have wanted to see it when it was still using long stringed together OR conditions!

I also think that several statuses could be removed, like the `LED state` and `IoT state` mentioned above, and possibly some more, but until someone who has the actual device joins in with some sample data and suggestions on the necessary properties I'm just leaving it as is for the time being, obviously without any intention on submitting it like that ;)

**BTW** with the recent conversions of some decoders to use the **bit** condition the required **m_docMax** has gone down for the first time; to 5895, but I didn't touch the current 7168 to leave some breathing space.
 
**ADDENDUM:** With the recent bit condition introduction it often is the case that two conditions for the same property have static results solely depending on the `0` or `1 `state of the bit. If some new static_value addition here would also reduce the size of several decoders, like

```
      "movement":{
         "condition":["servicedata", 2, "bit", 2, 0],
         "decoder":["static_value", false],
         "is_bool":1
      },
      "_movement":{
         "condition":["servicedata", 2, "bit", 2, 1],
         "decoder":["static_value", true],
         "is_bool":1
      },
```

to

```
      "movement":{
         "decoder":["static_value", "servicedata", 2, "bit", 2, ? false : true],
         "is_bool":1
      },
```

or

```
      "mode":{
         "condition":["servicedata", 2, "bit", 3, 1],
         "decoder":["static_value", "on/off"]
      },
      "_mode":{
         "condition":["servicedata", 2, "bit", 3, 0],
         "decoder":["static_value", "onestate"]
      },
      "state":{
         "condition":["servicedata", 2, "bit", 2, 0],
         "decoder":["static_value", "on"]
      },
      "_state":{
         "condition":["servicedata", 2, "bit", 2, 1],
         "decoder":["static_value", "off"]
      },

```
to
```
      "mode":{
         "decoder":["static_value", "servicedata", 2, "bit", 3, ? "onestate" : "on/off"]
      },
      "state":{
         "decoder":["static_value", "servicedata", 2, "bit", 2, ? "on" : "off"]
      },
```

in the SwitchBot Bot decoder - just a thought for possible future further compressions.

**Reply by DigiH** [reply]
Initial test with my lacking experience, but I think the compression in the example SwitchBot Bot, Xiaomi Sclaes and CGH1 decoders is quite desirable.

@h2zero I'm sure you have a few correction/improvement suggestions :)

https://github.com/DigiH/decoder/tree/static-bitwise-test

Even with my current, to be optimised, code

```
OMG [env:esp32dev-ble]

decoder = https://github.com/theengs/decoder.git#v0.4.0
RAM:   [==        ]  17.8% (used 58384 bytes from 327680 bytes)
Flash: [=======   ]  73.5% (used 1444280 bytes from 1966080 bytes)

decoder = https://github.com/DigiH/decoder#static-bitwise-test
RAM:   [==        ]  17.8% (used 58384 bytes from 327680 bytes)
Flash: [=======   ]  73.4% (used 1443496 bytes from 1966080 bytes)
```

at least a slight reduction with the so far converted decoders.

I suppose here we don't actually need the upper case evaluations, as servicedata and manufacturerdata are always present in lower case ...

Thanks

**Reply by DigiH** [reply]
• condensed getBinaryData to its own function, it being called twice
• commented out the UPPER case condition, as all presented data should only be presented lower case
• commented out the static_value bool cast which I introduced in https://github.com/theengs/decoder/pull/96, as this doesn't seem to be necessary any longer, possibly due to an included library change?!? Ok to delete it, or better to keep it in, just in case a future change might revert this again?

Thanks
